sstables/sstable_set.cc

```diff
@@ -168,10 +168,22 @@ void sstable_set::for_each_sstable(std::function<void(const shared_sstable&)> fu
     });
 }
 
+future<> sstable_set::for_each_sstable_gently(std::function<future<>(const shared_sstable&)> func) const {
+    return _impl->for_each_sstable_gently_until([func = std::move(func)] (const shared_sstable& sst) {
+        return func(sst).then([] {
+            return stop_iteration::no;
+        });
+    }).discard_result();
+}
```

Comment:
The sstable might not be interesting by the time we reach it.

It's really the caller's problem, but how to prevent doing work on an sstable that's already on the way to being discarded?

Also, if the sstable is being compacted, we might resurrect the sstable via cleanup or something.

Reply:
> Also, if the sstable is being compacted, we might resurrect the sstable via cleanup or something.

We do these traversals with compaction temporarily disabled to prevent this race.

> Also: coroutines?

can do